File: python/priorityOrderExample/watch_priority_mempool.py
#!/usr/bin/env python3
"""
Watch live priority-fee activity from a QuickNode Hyperliquid gRPC endpoint.

By default this example subscribes to the derived ORDER_PRIORITY stream and filters
for source=mempool_txs. These are pre-consensus mempool signals, not finalized
orders. Use them to see priority-fee order flow before it lands on-chain.
"""

import argparse
import json
import os
import sys
import time

import grpc
import zstandard as zstd

try:
    import hyperliquid_pb2 as pb
    import hyperliquid_pb2_grpc as pb_grpc
except ImportError:
    print("Error: Proto files not generated. Run:")
    print("  python -m grpc_tools.protoc -I../../proto --python_out=. --grpc_python_out=. ../../proto/hyperliquid.proto")
    sys.exit(1)


DEFAULT_GRPC_ENDPOINT = "your-endpoint.hype-mainnet.quiknode.pro:10000"
DEFAULT_AUTH_TOKEN = "YOUR_QUICKNODE_TOKEN"


def env_or_default(name: str, default: str) -> str:
    value = os.environ.get(name)
    return value if value else default


GRPC_ENDPOINT = env_or_default("GRPC_ENDPOINT", DEFAULT_GRPC_ENDPOINT)
AUTH_TOKEN = env_or_default("AUTH_TOKEN", env_or_default("QN_AUTH_TOKEN", DEFAULT_AUTH_TOKEN))
decompressor = zstd.ZstdDecompressor()
ZSTD_MAGIC = b"\x28\xB5\x2F\xFD"


def string_has_zstd_magic(data: str) -> bool:
    if len(data) < 4:
        return False
    try:
        return bytes(ord(char) for char in data[:4]) == ZSTD_MAGIC
    except ValueError:
        return False


def decompress(data) -> str:
    if isinstance(data, str):
        if string_has_zstd_magic(data):
            return decompressor.decompress(data.encode("latin-1")).decode("utf-8")
        return data
    if not data or len(data) < 4:
        return data.decode("utf-8") if isinstance(data, bytes) else str(data)
    if data[0:4] == ZSTD_MAGIC:
        return decompressor.decompress(data).decode("utf-8")
    return data.decode("utf-8")


def create_channel():
    return grpc.secure_channel(
        GRPC_ENDPOINT,
        grpc.ssl_channel_credentials(),
        options=[
            ("grpc.max_receive_message_length", 100 * 1024 * 1024),
            ("grpc.keepalive_time_ms", 30000),
            ("grpc.keepalive_timeout_ms", 10000),
        ],
    )


def request_generator(start_block: int, raw_mempool: bool, include_confirmed: bool):
    stream_type = "MEMPOOL_TXS" if raw_mempool else "ORDER_PRIORITY"
    filters = {}
    if not raw_mempool and not include_confirmed:
        filters["source"] = pb.FilterValues(values=["mempool_txs"])

    yield pb.SubscribeRequest(
        subscribe=pb.StreamSubscribe(
            stream_type=pb.StreamType.Value(stream_type),
            start_block=start_block,
            filters=filters,
        )
    )

    while True:
        time.sleep(30)
        yield pb.SubscribeRequest(ping=pb.Ping(timestamp=int(time.time() * 1000)))


def text_matches_filters(text: str, needles: list[str]) -> bool:
    return not needles or any(needle in text for needle in needles)


def priority_fees(value) -> list[str]:
    fees = []
    if isinstance(value, dict):
        if value.get("source") and value.get("type") == "order" and "p" in value:
            fees.append(str(value["p"]))
        grouping = value.get("grouping")
        if isinstance(grouping, dict) and "p" in grouping:
            fees.append(str(grouping["p"]))
        for item in value.values():
            fees.extend(priority_fees(item))
    elif isinstance(value, list):
        for item in value:
            fees.extend(priority_fees(item))
    return fees


def parse_json(text: str):
    try:
        return json.loads(text)
    except json.JSONDecodeError:
        return None


def print_payload(block_number: int, timestamp: int, text: str, compact: bool, fees: list[str]):
    print(f"\nBlock {block_number} | Timestamp {timestamp}")
    if fees:
        print(f"Priority p: {', '.join(fees)}")
    if compact:
        print(text[:1000])
        return

    parsed = parse_json(text)
    if parsed is not None:
        print(json.dumps(parsed, indent=2))
    else:
        print(text)


def main() -> int:
    parser = argparse.ArgumentParser(description="Watch pre-consensus Hyperliquid priority-fee mempool events from a QuickNode gRPC endpoint.")
    parser.add_argument("--start-block", type=int, default=0, help="Start block for the stream. Default 0.")
    parser.add_argument("--contains", action="append", default=[], help="Only print messages containing this text. Can be repeated.")
    parser.add_argument("--include-confirmed", action="store_true", help="Also include confirmed ORDER_PRIORITY events from replica_cmds.")
    parser.add_argument("--raw-mempool", action="store_true", help="Subscribe to raw MEMPOOL_TXS and detect grouping.p locally.")
    parser.add_argument("--all-mempool", action="store_true", help="With --raw-mempool, print all MEMPOOL_TXS messages, not only priority transactions.")
    parser.add_argument("--max-messages", type=int, help="Stop after printing this many matching messages.")
    parser.add_argument("--compact", action="store_true", help="Print compact payload text instead of pretty JSON.")
    args = parser.parse_args()
    raw_mempool = args.raw_mempool or args.all_mempool

    if GRPC_ENDPOINT == DEFAULT_GRPC_ENDPOINT:
        print("Set GRPC_ENDPOINT to your QuickNode Hyperliquid mainnet or testnet gRPC endpoint.", file=sys.stderr)
        return 2
    if AUTH_TOKEN == DEFAULT_AUTH_TOKEN:
        print("Set AUTH_TOKEN to your QuickNode token.", file=sys.stderr)
        return 2

    if raw_mempool:
        print("Watching raw MEMPOOL_TXS")
    elif args.include_confirmed:
        print("Watching ORDER_PRIORITY events from mempool_txs and replica_cmds")
    else:
        print("Watching pre-consensus ORDER_PRIORITY mempool events")
    print(f"Endpoint: {GRPC_ENDPOINT}")
    if not raw_mempool and not args.include_confirmed:
        print("Server filter: source=mempool_txs (not finalized)")
    elif raw_mempool and not args.all_mempool:
        print("Local filter: priority grouping only")
    if args.contains:
        print(f"Text filters: {args.contains}")

    channel = create_channel()
    stub = pb_grpc.StreamingStub(channel)
    metadata = [("x-token", AUTH_TOKEN)]
    printed = 0

    try:
        for response in stub.StreamData(request_generator(args.start_block, raw_mempool, args.include_confirmed), metadata=metadata):
            if response.HasField("pong"):
                continue
            if not response.HasField("data"):
                continue

            data = response.data
            text = decompress(data.data)
            if not text_matches_filters(text, args.contains):
                continue

            parsed = parse_json(text)
            fees = priority_fees(parsed) if parsed is not None else []
            if raw_mempool and not args.all_mempool and not fees:
                continue

            printed += 1
            print_payload(data.block_number, data.timestamp, text, args.compact, fees)
            if args.max_messages and printed >= args.max_messages:
                return 0
    except grpc.RpcError as exc:
        print(f"gRPC error: {exc.code()} - {exc.details()}", file=sys.stderr)
        return 1
    except KeyboardInterrupt:
        print("\nStopping stream...")
        return 0
    finally:
        channel.close()


if __name__ == "__main__":
    raise SystemExit(main())

File: javascript/priorityOrderExample/watch_priority_mempool.js
// Watch live priority-fee mempool events from a QuickNode Hyperliquid gRPC endpoint.
const grpc = require('@grpc/grpc-js');
const protoLoader = require('@grpc/proto-loader');
const zstd = require('@mongodb-js/zstd');
const path = require('path');

const DEFAULT_GRPC_ENDPOINT = 'your-endpoint.hype-mainnet.quiknode.pro:10000';
const DEFAULT_AUTH_TOKEN = 'YOUR_QUICKNODE_TOKEN';
const GRPC_ENDPOINT = process.env.GRPC_ENDPOINT || DEFAULT_GRPC_ENDPOINT;
const AUTH_TOKEN = process.env.AUTH_TOKEN || process.env.QN_AUTH_TOKEN || DEFAULT_AUTH_TOKEN;
const PROTO_PATH = path.join(__dirname, '..', '..', 'proto', 'hyperliquid.proto');

const packageDefinition = protoLoader.loadSync(PROTO_PATH, {
  keepCase: true,
  longs: String,
  enums: String,
  defaults: true,
  oneofs: true
});
const proto = grpc.loadPackageDefinition(packageDefinition).hyperliquid;

function parseArgs() {
  const args = process.argv.slice(2);
  const get = (name) => {
    const arg = args.find(a => a.startsWith(`--${name}=`));
    return arg ? arg.split('=').slice(1).join('=') : null;
  };

  const contains = [];
  const serverFilters = {};
  args.forEach((arg, i) => {
    if (arg === '--contains' && args[i + 1]) contains.push(args[i + 1]);
    if (arg.startsWith('--contains=')) contains.push(arg.split('=').slice(1).join('='));
    if (arg === '--filter' && args[i + 1]) addServerFilter(serverFilters, args[i + 1]);
    if (arg.startsWith('--filter=')) addServerFilter(serverFilters, arg.split('=').slice(1).join('='));
  });

  return {
    startBlock: parseInt(get('start-block') || '0', 10),
    contains,
    serverFilters,
    includeConfirmed: args.includes('--include-confirmed'),
    rawMempool: args.includes('--raw-mempool'),
    allMempool: args.includes('--all-mempool'),
    compact: args.includes('--compact'),
    maxMessages: get('max-messages') ? parseInt(get('max-messages'), 10) : null
  };
}

function addServerFilter(filters, expression) {
  const index = expression.indexOf('=');
  if (index <= 0 || index === expression.length - 1) {
    throw new Error(`Invalid --filter ${expression}; expected field=value1,value2`);
  }

  const field = expression.slice(0, index).trim();
  const values = expression
    .slice(index + 1)
    .split(',')
    .map(value => value.trim())
    .filter(Boolean);
  if (!field || values.length === 0) {
    throw new Error(`Invalid --filter ${expression}; expected field=value1,value2`);
  }
  filters[field] = { values };
}

async function decompress(data) {
  if (typeof data === 'string') {
    if (
      data.length >= 4 &&
      data.charCodeAt(0) === 0x28 &&
      data.charCodeAt(1) === 0xB5 &&
      data.charCodeAt(2) === 0x2F &&
      data.charCodeAt(3) === 0xFD
    ) {
      const out = await zstd.decompress(Buffer.from(data, 'latin1'));
      return out.toString('utf8');
    }
    return data;
  }
  if (!Buffer.isBuffer(data)) return String(data);
  if (data.length >= 4 && data[0] === 0x28 && data[1] === 0xB5 && data[2] === 0x2F && data[3] === 0xFD) {
    const out = await zstd.decompress(data);
    return out.toString('utf8');
  }
  return data.toString('utf8');
}

function priorityFees(value) {
  const fees = [];
  if (Array.isArray(value)) {
    value.forEach(item => fees.push(...priorityFees(item)));
  } else if (value && typeof value === 'object') {
    if (value.source && value.type === 'order' && value.p !== undefined) {
      fees.push(String(value.p));
    }
    if (value.grouping && typeof value.grouping === 'object' && value.grouping.p !== undefined) {
      fees.push(String(value.grouping.p));
    }
    Object.values(value).forEach(item => fees.push(...priorityFees(item)));
  }
  return fees;
}

function matchesTextFilters(text, contains) {
  return contains.length === 0 || contains.some(needle => text.includes(needle));
}

function createClient() {
  return new proto.Streaming(
    GRPC_ENDPOINT,
    grpc.credentials.createSsl(),
    {
      'grpc.max_receive_message_length': 100 * 1024 * 1024,
      'grpc.keepalive_time_ms': 30000,
      'grpc.keepalive_timeout_ms': 10000
    }
  );
}

function main() {
  const args = parseArgs();
  const rawMempool = args.rawMempool || args.allMempool;

  if (GRPC_ENDPOINT === DEFAULT_GRPC_ENDPOINT) {
    console.error('Set GRPC_ENDPOINT to your QuickNode Hyperliquid mainnet or testnet gRPC endpoint.');
    process.exit(2);
  }
  if (AUTH_TOKEN === DEFAULT_AUTH_TOKEN) {
    console.error('Set AUTH_TOKEN to your QuickNode token.');
    process.exit(2);
  }

  if (rawMempool) {
    console.log('Watching raw MEMPOOL_TXS');
  } else if (args.includeConfirmed) {
    console.log('Watching ORDER_PRIORITY events from mempool_txs and replica_cmds');
  } else {
    console.log('Watching pre-consensus ORDER_PRIORITY mempool events');
  }
  console.log(`Endpoint: ${GRPC_ENDPOINT}`);
  if (!rawMempool && !args.includeConfirmed) {
    console.log('Server filter: source=mempool_txs (not finalized)');
  } else if (rawMempool && !args.allMempool) {
    console.log('Local filter: priority grouping only');
  }
  if (Object.keys(args.serverFilters).length) {
    console.log(`Server filters: ${JSON.stringify(args.serverFilters)}`);
  }
  if (args.contains.length) console.log(`Text filters: ${JSON.stringify(args.contains)}`);

  const client = createClient();
  const metadata = new grpc.Metadata();
  metadata.add('x-token', AUTH_TOKEN);
  const call = client.StreamData(metadata);
  let printed = 0;
  let stopping = false;
  let processing = Promise.resolve();
  const filters = !rawMempool && !args.includeConfirmed
    ? { source: { values: ['mempool_txs'] }, ...args.serverFilters }
    : args.serverFilters;

  call.write({
    subscribe: {
      stream_type: rawMempool ? 'MEMPOOL_TXS' : 'ORDER_PRIORITY',
      start_block: args.startBlock,
      filters
    }
  });

  const ping = setInterval(() => {
    call.write({ ping: { timestamp: Date.now() } });
  }, 30000);

  async function handleData(response) {
    if (stopping || !response.data) return;

    let text;
    try {
      text = await decompress(response.data.data);
    } catch (err) {
      console.warn(`Decompress error at block ${response.data.block_number}: ${err.message}`);
      return;
    }
    if (!matchesTextFilters(text, args.contains)) return;

    let parsed = null;
    try {
      parsed = JSON.parse(text);
    } catch {}

    const fees = parsed ? priorityFees(parsed) : [];
    if (rawMempool && !args.allMempool && fees.length === 0) return;
    if (args.maxMessages && printed >= args.maxMessages) return;

    printed += 1;
    console.log(`\nBlock ${response.data.block_number} | Timestamp ${response.data.timestamp}`);
    if (fees.length) console.log(`Priority p: ${fees.join(', ')}`);
    if (args.compact) {
      console.log(text.slice(0, 1000));
    } else if (parsed) {
      console.log(JSON.stringify(parsed, null, 2));
    } else {
      console.log(text);
    }

    if (args.maxMessages && printed >= args.maxMessages) {
      stopping = true;
      clearInterval(ping);
      call.end();
    }
  }

  call.on('data', (response) => {
    processing = processing
      .then(() => handleData(response))
      .catch((err) => {
        console.error(`Handler error: ${err.message}`);
      });
  });

  call.on('error', (err) => {
    clearInterval(ping);
    console.error(`gRPC error: ${err.message}`);
    process.exit(1);
  });

  call.on('end', () => {
    clearInterval(ping);
  });
}

main();
